graphql/schemabuilder/helpers.go

```diff
@@ -0,0 +1,626 @@
+
+	in := make([]reflect.Type, 0, funcType.NumIn())
+	for i := 0; i < funcType.NumIn(); i++ {
+		in = append(in, funcType.In(i))
+	}
+
+	var argParser *argParser
+	var argType graphql.Type
+	var ptrFunc bool
+	var hasContext, hasSource, hasSelectionSet bool
+
+	if len(in) > 0 && in[0] == contextType {
+		hasContext = true
+		in = in[1:]
+	}
+	if len(in) > 0 && (in[0] == typ || in[0] == ptr) {
+		hasSource = true
+		ptrFunc = in[0] == ptr
+		in = in[1:]
+	}
+
+	if len(in) > 0 && in[0] != selectionSetType {
+		var err error
+		if argParser, argType, err = sb.buildPaginatedArgParser(in[0]); err != nil {
+			return nil, fmt.Errorf("attempted to wrap %s as arguments struct, but failed: %s", in[0].Name(), err.Error())
+		}
+		in = in[1:]
+	} else {
+		var err error
+		if argParser, argType, err = sb.buildPaginatedArgParser(nil); err != nil {
+			return nil, fmt.Errorf("test this case...shouldn't be failing in any case")
+		}
+	}
+
+	if len(in) > 0 && in[0] == selectionSetType {
+		hasSelectionSet = true
+		in = in[:len(in)-1]
+	}
+
+	// We have succeeded if no arguments remain.
+	if len(in) != 0 {
+		return nil, fmt.Errorf("%s arguments should be [context][, [*]%s][, args][, selectionSet]", funcType, typ)
+	}
+
+	// Parse return values. The first return value must be the actual value, and
+	// the second value can optionally be an error.
+
+	out := make([]reflect.Type, 0, funcType.NumOut())
+	for i := 0; i < funcType.NumOut(); i++ {
+		out = append(out, funcType.Out(i))
+	}
+
+	var hasRet, hasError bool
+
+	var nodeType reflect.Type
+	if len(out) > 0 && out[0] != errType {
+		nodeType = out[0].Elem()
+		hasRet = true
+		out = out[1:]
+	} else {
+		return nil, fmt.Errorf("paginated function must have a return argument")
+	}
+
+	if len(out) > 0 && out[0] == errType {
+		hasError = true
+		out = out[1:]
+	}
+
+	if len(out) != 0 {
+		return nil, fmt.Errorf("%s return values should [result][, error]", funcType)
+	}
+
+	if !hasRet {
+		return nil, fmt.Errorf("%s has no return value", funcType)
+	}
+
+	var retType graphql.Type
+	if hasRet {
+		var err error
+		retType, err = sb.constructConnType(nodeType)
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		var err error
+		retType, err = sb.getType(reflect.TypeOf(true))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	args := make(map[string]graphql.Type)
+
+	inputObject, ok := argType.(*graphql.InputObject)
+	if !ok {
+		return nil, fmt.Errorf("%s's args should be an object", funcType)
+	}
+
+	for name, nodeType := range inputObject.InputFields {
+		args[name] = nodeType
+	}
+
+	ret := &graphql.Field{
+		Resolve: func(ctx context.Context, source, args interface{}, selectionSet *graphql.SelectionSet) (interface{}, error) {
+
+			if _, ok := args.(ConnectionArgs); !ok {
+				return nil, fmt.Errorf("arguments should implement ConnectionArgs")
+			}
+
+			in := make([]reflect.Value, 0, funcType.NumIn())
+
+			if hasContext {
+				in = append(in, reflect.ValueOf(ctx))
+			}
+
+			// Set up source.
```

Comment:
can we de-dupe any of the resolve function with the other implementation?

Reply:
I think we can remove a lot of boilerplate in these helper functions so it might not be worth refactoring reflect code